src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+
+            return references;
+        }
+
+        private void GetProjectSnapshotAsync(ConfiguredProject selectedConfiguredProject)
+        {
+            _projectReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _packageReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _assemblyReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _sdkReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+        }
+
+        private List<ProjectSystemReferenceInfo> GetReferences()
+        {
+            List<ProjectSystemReferenceInfo> references = new List<ProjectSystemReferenceInfo>();
+
+            references.AddRange(_projectReferenceHandler.GetReferences());
+            references.AddRange(_packageReferenceHandler.GetReferences());
+            references.AddRange(_assemblyReferenceHandler.GetReferences());
+            references.AddRange(_sdkReferenceHandler.GetReferences());
+
+            return references;
+        }
+
+        public async Task<bool> TryUpdateReferenceAsync(string projectPath, string targetFrameworkMoniker, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+        {
+            bool wasUpdated = false;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.None)
+            {
+                return wasUpdated;
+            }
+
+            var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+            var referenceHandler = FindReferenceHandler(referenceUpdate);
+
+            wasUpdated = await ApplyAction(referenceUpdate, cancellationToken, referenceHandler, activeConfiguredProject);
+
+            return wasUpdated;
+        }
+
+        private static async Task<bool> ApplyAction(ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken,
+            ReferenceHandler referenceHandler, ConfiguredProject activeConfiguredProject)
+        {
+            bool wasUpdated;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ||
+                referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUnused)
+            {
+                wasUpdated =
+                    await referenceHandler.UpdateReferenceAsync(activeConfiguredProject, referenceUpdate, cancellationToken);
+            }
+            else
+            {
+                if (!canRemoveReference(referenceUpdate, referenceHandler))
+                {
+                    wasUpdated = false;
+                }
+                else
+                {
+                    wasUpdated =
+                        await referenceHandler.RemoveReferenceAsync(activeConfiguredProject, referenceUpdate.ReferenceInfo);
+                }
+            }
+
+            return wasUpdated;
+        }
+
+        private static bool canRemoveReference(ProjectSystemReferenceUpdate referenceUpdate, ReferenceHandler referenceHandler)
```

Comment:
**C**anRemoveReference

Reply:
Can we add a warning rule for methods to be Pascal ? I added this, but didn't detect anything.

![image](https://user-images.githubusercontent.com/510598/96894901-c295ed00-1440-11eb-997f-e8c7a00569d5.png)